flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/source/enumerator/tracker/SplitTracker.java

```diff
@@ -89,13 +91,77 @@ public SplitTracker(
      * @param shardsToAdd collection of splits to add to tracking
      */
     public void addSplits(Collection<Shard> shardsToAdd) {
-        Set<String> discoveredShardIds =
-                shardsToAdd.stream().map(Shard::shardId).collect(Collectors.toSet());
+        if (TRIM_HORIZON.equals(initialPosition)) {
+            addSplitsForTrimHorizon(shardsToAdd);
+            return;
+        }
+        addSplitsForLatest(shardsToAdd);
+    }
+
+    /**
+     * If there is an ancestor of an open shard in knownSplits, we put all of its ancestors after
+     * the tracked ancestors as TRIM_HORIZON otherwise, we ignore all the ancestors and track only
+     * the open shard with LATEST. There might be a case when restoring from expired snapshots that
+     * there is an ancestor in knownSplits, but we wont know that it is an ancestor, so in that case
+     * also, all the ancestors will not be tracked and rather only the current open shard will be
+     * tracked with LATEST.
```

Comment:
I don't think this is correct. When restoring from snapshot, we should read all records since the restored snapshot (`TRIM_HORIZON`), and not `LATEST`.

Reply:
We should align that the expected behavior, when using `LATEST`, should be:

1. When starting without snapshot,
   - All shards read from `LATEST`. If we want, we can focus on just the open shards.
2. When reading with snapshot,
   - All shards from time of start of application are read from `TRIM_HORIZON`,

Reply:
changed the algo to resolve this